### ISSUE TITLE
Document toolchain setup for Windows

### ISSUE THREAD

**Issue by str4d** [Ralith/openxrs#46]
I've been attempting to run the `vulkan` example on Windows. This has taken a while to figure out:

- Straight out of the box, the build fails (somewhat expected). The `OpenXR-SDK` `README` mentions that `msbuild.exe` needs to be in the `PATH`, but it's unclear whether this alone is sufficient.
- The Visual Studio 2019 documentation points me towards the "Developer PowerShell for VS 2019" shortcut, so I try using that to open a shell in which the CLI has the necessary tooling available. This unfortunately only supports 32-bit tooling, and the 64-bit tooling uses `cmd.exe` which is terrible, but [this community thread](https://developercommunity.visualstudio.com/idea/943058/x64-developer-powershell-for-vs-2019.html) has the necessary flags for creating a 64-bit PowerShell environment. I create a PowerShell script for this because I run into character limits while trying to modify the default shortcut.

- In the VS 2019 64-bit PowerShell environment, `openxr-sys` still fails to build, with the following error:

```
error: failed to run custom build command for `openxr-sys v0.8.4 (C:\Users\me\Documents\dev\rust\sat\openxrs\sys)`

Caused by:
  process didn't exit successfully: `C:\Users\me\Documents\dev\rust\sat\openxrs\target\release\build\openxr-sys-646a411395f24767\build-script-build` (exit code: 101)
--- stdout
running: "cmake" "C:\\Users\\me\\Documents\\dev\\rust\\sat\\openxrs\\sys\\OpenXR-SDK" "-G" "Visual Studio 16 2019" "-Thost=x64" "-Ax64" "-DBUILD_API_LAYERS=OFF" "-DBUILD_TESTS=OFF" "-DOpenGL_GL_PREFERENCE=GLVND" "-DDYNAMIC_LOADER=OFF" "-DCMAKE_INSTALL_LIBDIR=lib" "-DCMAKE_INSTALL_PREFIX=C:\\Users\\me\\Documents\\dev\\rust\\sat\\openxrs\\target\\release\\build\\openxr-sys-72146758c318d865\\out" "-DCMAKE_C_FLAGS= -nologo -MD -Brepro" "-DCMAKE_C_FLAGS_RELEASE= -nologo -MD -Brepro" "-DCMAKE_CXX_FLAGS= -nologo -MD -Brepro" "-DCMAKE_CXX_FLAGS_RELEASE= -nologo -MD -Brepro" "-DCMAKE_ASM_FLAGS= -nologo -MD -Brepro" "-DCMAKE_ASM_FLAGS_RELEASE= -nologo -MD -Brepro" "-DCMAKE_BUILD_TYPE=Release"
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.19041.
-- The C compiler identification is MSVC 19.27.29111.0
-- The CXX compiler identification is MSVC 19.27.29111.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: C:/Users/me/AppData/Local/Programs/Python/Python38-32/python.exe (found suitable version "3.8.1", minimum required is "3")
-- Found OpenGL: opengl32
-- Enabling OpenGL support
-- Could NOT find VulkanHeaders (missing: VulkanHeaders_INCLUDE_DIR)
-- Could NOT find VulkanRegistry (missing: VulkanRegistry_DIR)
-- Could NOT find Vulkan (missing: Vulkan_LIBRARY Vulkan_INCLUDE_DIR)
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE
-- Could NOT find JsonCpp (missing: JsonCpp_INCLUDE_DIR JsonCpp_LIBRARY)
-- Could NOT find glslc, using precompiled .spv files
-- Looking for secure_getenv
-- Looking for secure_getenv - not found
-- Looking for __secure_getenv
-- Looking for __secure_getenv - not found
-- Looking for timespec_get
-- Looking for timespec_get - found
-- Found and will use pre-generated xr_generated_dispatch_table.h in source tree
-- Found and will use pre-generated xr_generated_dispatch_table.c in source tree
-- Found and will use pre-generated xr_generated_loader.hpp in source tree
-- Found and will use pre-generated xr_generated_loader.cpp in source tree
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/me/Documents/dev/rust/sat/openxrs/target/release/build/openxr-sys-72146758c318d865/out/build
running: "cmake" "--build" "." "--target" "install" "--config" "Release" "--"
Microsoft (R) Build Engine version 16.7.0+b89cb5fde for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

MSBUILD : error MSB1009: Project file does not exist.
Switch: install.vcxproj

--- stderr
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_ASM_FLAGS
    CMAKE_ASM_FLAGS_RELEASE
    CMAKE_BUILD_TYPE


thread 'main' panicked at '
command did not execute successfully, got: exit code: 1

build script failed, must exit now', C:\Users\me\.cargo\registry\src\github.com-1ecc6299db9ec823\cmake-0.1.44\src\lib.rs:885:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

warning: build failed, waiting for other jobs to finish...
error: build failed
```

It turns out that this is because the file being generated is `INSTALL.vcxproj`, and either `CMake` or `MSBUILD` is case-sensitive. Fortunately, renaming the generated file under `target/release/build/openxr-sys-../out/build` is sufficient to get past this.
- I then discover that the MSRV for the crate is greater than 1.40.0 (which is what I had installed at the time) due to these errors:
```
error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
   --> sys\src\lib.rs:191:6
    |
191 | impl<T> std::ops::Index<HandJointEXT> for [T] {
    |      ^ type parameter `T` must be used as the type parameter for some local type
    |
    = note: only traits defined in the current crate can be implemented for a type parameter

error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
   --> sys\src\lib.rs:198:6
    |
198 | impl<T> std::ops::IndexMut<HandJointEXT> for [T] {
    |      ^ type parameter `T` must be used as the type parameter for some local type
    |
    = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to 2 previous errors
```
and the MSRV for the example is at least 1.43.0 due to the use of `u64::MAX`.
- After updating to Rust 1.46.0, compiling to hit the case-sensitive bug, renaming the file, and compiling again, I finally can run `vulcan.exe`, and it crashes:
```
     Running `C:\Users\me\Documents\dev\rust\sat\openxrs\target\debug\examples\vulkan.exe`
loaded OpenXR runtime: Oculus 1.52.0
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NulError(38, [86, 75, 95, 75, 72, 82, 95, 103, 101, 116, 95, 112, 104, 121, 115, 105, 99, 97, 108, 95, 100, 101, 118, 105, 99, 101, 95, 112, 114, 111, 112, 101, 114, 116, 105, 101, 115, 50, 0])', openxr\examples\vulkan.rs:65:34
```
which appears to be failing because `CString::new` (which appends a null byte) is being called on the already-null-terminated string `VK_KHR_get_physical_device_properties2\x00`:
https://github.com/Ralith/openxrs/blob/b7eed859fbeeb1b44d77a416c627b90be565d86f/openxr/examples/vulkan.rs#L61-L65

There's clearly several steps in the toolchain setup process that I'm missing; it would be great to have them documented.

**Comment by Ralith** [comment]
Thanks for the issue, and for your patience working through all those surprises!

> It turns out that this is because the file being generated is INSTALL.vcxproj, and either CMake or MSBUILD is case-sensitive.

This sounds like an upstream build system bug. I recommend opening an issue at https://github.com/KhronosGroup/OpenXR-SDK-Source, assuming it can be replicated by hand. I'm not sure why I haven't encountered this myself; I'm using the VS 2017 build tools, and for environment setup I invoke the usual .bat script and then launch powershell afterwards. In any case, I don't mind adding a brief aside regarding windows dev env setup to the README. Maybe we should just link the MSVC docs? Which did you read?

> MSRV

Apologies for the surprises there! I'd happily accept a PR to document and test a consistent MSRV.

> `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NulError`

This looks like an Oculus bug; discusson continues in #47.

**Comment by str4d** [comment]
> I'm using the VS 2017 build tools, and for environment setup I invoke the usual .bat script and then launch powershell afterwards.

Heh, that's much simpler 😄 Running the "x64 Native Tools Command Prompt" (which runs `vcvars64.bat`) and then `powershell` within it does get me to a working environment. I still encounter the case-sensitive bug; maybe it's specific to VS 2019?

> Which did you read?

I read these docs: https://docs.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=vs-2019#developer_command_prompt_shortcuts

**Comment by Callum17** [comment]
Just to muddy the waters: I've been building on windows via the standard powershell env without any real issue. 
Didn't have to do any extra setup apart from enabling the static feature in the openxr package:
```
diff --git a/openxr/Cargo.toml b/openxr/Cargo.toml
-default = ["loaded"]
+default = ["loaded", "static"]
```

Worth mentioning I did run into some minor runtime compatibility issues: #68

Setup:
visual studio 2019
rustc 1.49.0 (stable)
cmake 3.19.3
Windows 10 Home
Quest 2 -> virtual desktop -> steamVR runtime

For me `get-command msbuild` showed `msbuild.exe` is coming from C:\Windows\Microsoft.NET\Framework\v4.0.30319.
Which is itself in `Path` as shown by `$Env:Path`.

`msbuild.exe` returned `4.8.4084.0` which interestingly is an [older version for vscode 2012](https://en.wikipedia.org/wiki/MSBuild). Not entirely sure how much the msbuild version matters, or when this was added to my `Path`...
Was building without issue at this point: `cargo build --release --examples`

Followed up by adding `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin` to `Path` via control panel -> system -> edit the sytem envrionment variables.

Confirmed `msbuild.exe` returned version 16.X for vs2019 `16.7.0+b89cb5fde`. Built again without issue: `cargo build --release --examples`.
